### PR TITLE
Issue746

### PR DESCRIPTION
**Pull request by loay**

### CODE REVIEW

File: lib/passport-configurator.js [hunk 1]
```diff
@@ -294,31 +294,32 @@ PassportConfigurator.prototype.configureProvider = function(name, options) {
                     function(err, accessToken) {
                       if (err) {
                         return err.code === 'LOGIN_FAILED'
-                            ? done(null, false, {message: 'Login failed'})
+                            ? done(null, false, {message: errorMsg})
```
Comment (superkhau) [inline lib/passport-configurator.js]: Is there a reason to make the error messages generic? The ones already in place seem to identify the problem more specifically.

File: lib/passport-configurator.js [hunk 2]
```diff
@@ -294,31 +294,32 @@ PassportConfigurator.prototype.configureProvider = function(name, options) {
...
                             : done(err);
                       }
-                      if (accessToken) {
+                      if (accessToken && user.emailVerified) {
                         userProfile.accessToken = accessToken;
                         done(null, userProfile, {accessToken: accessToken});
                       } else {
-                        done(null, false, {message: 'Failed to create token.'});
+                        done(null, false, {message: errorMsg});
```
Comment (superkhau) [inline lib/passport-configurator.js]: ditto

File: lib/passport-configurator.js [hunk 3]
```diff
@@ -294,31 +294,32 @@ PassportConfigurator.prototype.configureProvider = function(name, options) {
...
                       }
                     });
                 }
-              }
+              } else {
```
Comment (superkhau) [inline lib/passport-configurator.js]: why add else, line 310 has a return in front to short circuit

File: test/model.user-identity.test.js [hunk 4]
```diff
@@ -187,9 +187,53 @@ describe('UserIdentity', function () {
...
           assert.equal(user.username, 'ldap.xyzldap');
           assert.equal(user.email, 'fooldap@bar.com');
-          
+
           done();
         });
       });
   });
+
+  it('supports accessToken to be true with local accounts', function (done) {
```
Comment (superkhau) [inline test/model.user-identity.test.js]: what does this message mean? describe the problem in a context without details to the implementation

File: test/model.user-identity.test.js [hunk 5]
```diff
@@ -187,9 +187,53 @@ describe('UserIdentity', function () {
...
+          assert.equal(user.id, 100);
+          assert.equal(user.emailVerified, false);
+            User.login({email: user.username, password: 'jdskdhlskh'}, function(err,user){
+              assert(err.code,  'LOGIN_FAILED');
+            });
+            User.login({email: user.username, password: 'pass'}, function(err,user){
+              assert(err.code,  'LOGIN_FAILED');
+            });
+          done();
+        });
```
Comment (superkhau) [inline test/model.user-identity.test.js]: tabbing is off here

File: test/model.user-identity.test.js [hunk 6]
```diff
@@ -187,9 +187,53 @@ describe('UserIdentity', function () {
...
+});
+
+  it('supports accessToken to be true with local accounts when email is verified', function (done) {
```
Comment (superkhau) [inline test/model.user-identity.test.js]: same for the message here, describe problem without impl details

File: test/model.user-identity.test.js [hunk 7]
```diff
@@ -187,9 +187,53 @@ describe('UserIdentity', function () {
...
+        assert.equal(user.emailVerified, true);
+          User.login({email: user.username, password: 'notpass'}, function(err,user){
+            assert(err.code,  'LOGIN_FAILED');
+          });
+          User.login({email: user.username, password: 'pass'}, function(err,user){
+            assert(!err,  'No error should be reported');
+          });
+        done();
+      });
+    });
```
Comment (superkhau) [inline test/model.user-identity.test.js]: tabbing

File: test/model.user-identity.test.js [hunk 8]
```diff
@@ -187,9 +187,53 @@ describe('UserIdentity', function () {
...
+    id: 101
+    }, function (err, user) {
+        assert.equal(user.username, 'myLocalTest2');
+        assert.equal(user.email, 'myLocalTest2@local.com');
+        assert.equal(user.id, 101);
+        assert.equal(user.emailVerified, true);
+          User.login({email: user.username, password: 'notpass'}, function(err,user){
+            assert(err.code,  'LOGIN_FAILED');
+          });
+          User.login({email: user.username, password: 'pass'}, function(err,user){
```
Comment (superkhau) [inline test/model.user-identity.test.js]: space after ,

**Comment by superkhau** [comment]
Make sure to lint your changes (if the linter is not impl'd yet, impl it and rebase). Some minor comments on the test case descriptions too.

File: lib/passport-configurator.js (hunk 1 above)
Comment (loay) [inline lib/passport-configurator.js]: Basically, it is about not giving out the root of the issue for security reasons, especially in production. We did the same thing last time with the username/password/verification errors. 
The error code is still the same, which can tell the root case. I don't mind reverting back to the main error message. The unit test still catches the error code.

File: lib/passport-configurator.js (hunk 3 above)
Comment (loay) [inline lib/passport-configurator.js]: that `else` is the main fix for the bug. if we remove it, we will get the error message of ```Error: Can't set headers after they are sent.```

**Comment by loay** [comment]
@superkhau PTAL. thanks.

File: test/model.user-identity.test.js [hunk 9]
```diff
@@ -187,9 +187,57 @@ describe('UserIdentity', function () {
...
   });
+
+  it('supports accessToken option before email verification', function(done) {
+    User.create({
+      username: 'myLocalTest',
+      email: 'myLocalTest@local.com',
+      password: 'pass',
+      emailVerified: 0,
+      id: 100
+    }, function (err, user) {
```
Comment (superkhau) [inline test/model.user-identity.test.js]: remove space after function

File: lib/passport-configurator.js (hunk 3 above)
Comment (superkhau) [inline lib/passport-configurator.js]: This is fine. It's the actual cause of the bug I see.

File: test/model.user-identity.test.js [hunk 10]
```diff
@@ -187,9 +187,57 @@ describe('UserIdentity', function () {
...
+      password: 'pass',
+      emailVerified: 1,
+      id: 101
+    }, function (err, user) {
+      assert.equal(user.username, 'myLocalTest2');
+      assert.equal(user.email, 'myLocalTest2@local.com');
+      assert.equal(user.id, 101);
+      assert.equal(user.emailVerified, true);
+      User.login({username: user.username, password: 'notpass'},
+        function (err, user) {
```
Comment (superkhau) [inline test/model.user-identity.test.js]: here too

File: test/model.user-identity.test.js [hunk 11]
```diff
@@ -187,9 +187,57 @@ describe('UserIdentity', function () {
...
+          assert(err, 'LOGIN_FAILED');
+        });
+      User.login({username: user.username, password: 'pass'},
+        function (err, user) {
```
Comment (superkhau) [inline test/model.user-identity.test.js]: ditto

File: test/model.user-identity.test.js [hunk 12]
```diff
@@ -187,9 +187,57 @@ describe('UserIdentity', function () {
...
+          if (err) return done(err);
+        });
+        done();
```
Comment (superkhau) [inline test/model.user-identity.test.js]: is this done in the wrong spot?

**Comment by superkhau** [comment]
Minor comments, LGTM. ;)

File: package.json [hunk 13]
```diff
@@ -28,7 +28,9 @@
   },
   "dependencies": {
     "passport": "^0.3.2",
-    "underscore": "^1.8.2"
+    "underscore": "^1.8.2",
+    "eslint": "^2.5.3",
+    "eslint-config-loopback": "^1.0.0"
   },
   "devDependencies": {
     "loopback": "^2.14.0",
```
Comment (superkhau) [inline package.json]: In the scripts section, add `"lint": "eslint ."` and also `"posttest": "npm run lint"`

**Comment by superkhau** [comment]
Please add the lint and posttest run tasks to the scripts section. Make sure it looks like:

```
"scripts": {
  "lint": "eslint .",
  "test": "...",
  "posttest": "npm run lint"
}
```

**Pull request merged**